pkg/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,177 @@
+
+	consumers, err := d.consumerDao.All(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to list consumers: %s", err.Error())
+	}
+
+	// initialize the consumer set for current instance
+	for _, consumer := range consumers {
+		instanceID := d.consistent.LocateKey([]byte(consumer.ID)).String()
+		if instanceID == d.instanceID {
+			d.consumerSet.Add(consumer.ID)
+			logger.Infof("Added new consumer %s to consumer set for instance %s", consumer.ID, d.instanceID)
+			d.resyncChan <- consumer.ID
+		}
+	}
+	logger.Infof("Initialized consumers %s for current instance %s", d.consumerSet.String(), d.instanceID)
+
+	pulseTicker := time.NewTicker(time.Duration(d.pulseInterval) * time.Second)
+	checkTicker := time.NewTicker(time.Duration(d.checkInterval) * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			pulseTicker.Stop()
+			checkTicker.Stop()
+			close(d.resyncChan)
+			return nil
+		case <-pulseTicker.C:
+			logger.Infof("Updating heartbeat for maestro instance: %s", instance.Name)
+			// update the heartbeat for current instance
+			instance.UpdatedAt = time.Now()
+			instance, err = d.instanceDao.Replace(ctx, instance)
+			if err != nil {
+				return fmt.Errorf("unable to update heartbeat for maestro instance: %s", err.Error())
+			}
+		case <-checkTicker.C:
+			logger.Infof("Checking maestro instances and updating the hash ring")
+			// check for instances that are not updated in the last instanceExpirationPeriod seconds
+			instances, err = d.instanceDao.All(ctx)
+			if err != nil {
+				return fmt.Errorf("unable to list maestro instances: %s", err.Error())
+			}
+			for _, instance := range instances {
+				if instance.UpdatedAt.After(time.Now().Add(time.Duration(-1*d.instanceExpirationPeriod) * time.Second)) {
+					d.consistent.Add(instance)
+				} else {
+					d.consistent.Remove(instance.Name)
+				}
+			}
+			logger.Infof("members in consistent hash ring: %v", d.consistent.GetMembers())
+			consumers, err = d.consumerDao.All(ctx)
+			if err != nil {
+				return fmt.Errorf("unable to list consumers: %s", err.Error())
+			}
+			for _, consumer := range consumers {
+				instanceID := d.consistent.LocateKey([]byte(consumer.ID)).String()
+				if instanceID == d.instanceID {
+					if added := d.consumerSet.Add(consumer.ID); added {
+						logger.Infof("Added new consumer %s to consumer set for instance %s", consumer.ID, d.instanceID)
+						d.resyncChan <- consumer.ID
+					}
+				} else {
+					// consistent hashing minimize this case but it is still possible
+					// remove the consumer from the set if it is not in the current instance
+					if d.consumerSet.Contains(consumer.ID) {
+						logger.Infof("Removed consumer %s from consumer set for instance %s", consumer.ID, d.instanceID)
+						d.consumerSet.Remove(consumer.ID)
+					}
+				}
+			}
+			logger.Infof("Consumers %s for current instance %s", d.consumerSet.String(), d.instanceID)
+		}
+	}
+}
+
+func (d *DispatcherImpl) Dispatch(consumerID string) bool {
```

Comment:
we may need a lock for consumerSet? or consumerSet is a threadsafe?

Reply:
the set we use is thread safe, see: https://github.com/deckarep/golang-set/?tab=readme-ov-file#features